File: Programmers/Lv2/target_number.cpp
#include <string>
#include <vector>

using namespace std;

void dfs(vector<int> number, int &ans, int target, int cnt = 0, int sum = 0)
{

    if (cnt == number.size() - 1)
    {
        if (sum + number[cnt] == target)
            ans++;
        if (sum - number[cnt] == target)
            ans++;
        return;
    }
    dfs(number, ans, target, cnt + 1, sum + number[cnt]);
    dfs(number, ans, target, cnt + 1, sum - number[cnt]);
}

int solution(vector<int> numbers, int target)
{
    int answer = 0;

    dfs(numbers, answer, target);

    return answer;
}

File: java/src/programmers/lv2/TargetNumber.java
package programmers.lv2;

public class TargetNumber {
    static int answer = 0;
    public int solution(int[] numbers, int target) {

        dfs(target,0,0,numbers);
        return answer;
    }
    void dfs(int target,int sum,int index, int[] numbers){
        if(index == numbers.length){
            if(sum == target)answer++;
            return;
        }
        dfs(target,sum+numbers[index], index+1, numbers);
        dfs(target,sum-numbers[index], index+1, numbers);
    }
}
